### PR TITLE
8317993: Add capturing factories to classes in java.util.function package

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to add a number of "capturing factories" in classes in the `java.util.function` package.

The PR additionally (an optionally) proposes to add a new function `UnaryOperator::andThenUnary` to allow composition while retaining the `UnaryOperator` type.

With the new changes, it is possible to write code like this (example from `java.util.function.Function`):

```
// Resolve ambiguity
var function = Function.of(String::isEmpty); // Function<String, Boolean>
var predicate = Predicate.of(String::isEmpty); // Predicate<String>

// Fluent composition
var chained = Function.of(String::length)       // Function<String, Integer>
                   .andThen(Integer::byteValue);      // Function<String, Byte>
```

Please see the original bug report for a comprehensive description of these proposed changes.

Note: It is not the objective to promote `var` declaration or to prevent previous ways of capturing lambdas and method references. The comments in the code above is for explaining the binding and once that becomes obvious, such comments are likely to not appear in real code. Users that prefers having a normal type declaration can still do that.

Note: Functional interfaces of primitives have not been considered (in this round). Otherwise, functional interfaces that might be ambiguous or that supports composition have been included. Hence, `Supplier` did not get a factory method.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a CSR request matching fixVersion 22 to be approved (needs to be created)

### Issues
 * [JDK-8317993](https://bugs.openjdk.org/browse/JDK-8317993): Add capturing factories to classes in java.util.function package (**Enhancement** - P4) ⚠️ Issue is not open.
 * [JDK-8318298](https://bugs.openjdk.org/browse/JDK-8318298): Add capturing factories to classes in java.util.function package (**CSR**) (Withdrawn)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16213/head:pull/16213` \
`$ git checkout pull/16213`

Update a local copy of the PR: \
`$ git checkout pull/16213` \
`$ git pull https://git.openjdk.org/jdk.git pull/16213/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16213`

View PR using the GUI difftool: \
`$ git pr show -t 16213`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16213.diff">https://git.openjdk.org/jdk/pull/16213.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16213#issuecomment-1765896818)

### CODE REVIEW

File: src/java.base/share/classes/java/util/function/UnaryOperator.java [hunk 1]
```diff
@@ -49,4 +51,46 @@ public interface UnaryOperator<T> extends Function<T, T> {
...
+     * the caller of the composed function.
+     *
+     * @param after the unary operator to apply after this unary operator is applied
+     * @return a composed unary operator that first applies this function and then
+     * applies the {@code after} unary operator
+     * @throws NullPointerException if after is null
+     *
+     * @see Function#andThen(Function)
+     */
+    default UnaryOperator<T> andThenUnary(UnaryOperator<T> after) {
```
Comment (minborg) [inline src/java.base/share/classes/java/util/function/UnaryOperator.java:68]: What are your comments on this proposed method?

File: src/java.base/share/classes/java/util/function/BiConsumer.java [hunk 2]
```diff
@@ -72,4 +72,37 @@ default BiConsumer<T, U> andThen(BiConsumer<? super T, ? super U> after) {
...
+     * var composed = BiConsumer.of(this::toConsole)
+     *                    .andThen(this::toLogger);  // BiConsumer<Long, String>
+     *
+     * }
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the first argument to the operation
+     * @param <U> the type of the second argument to the operation
+     */
+    static <T, U> BiConsumer<T, U> of(BiConsumer<T, U> uncaptured) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/function/BiConsumer.java:104]:
```suggestion
    static <T, U> BiConsumer<T, U> of(BiConsumer<? super T, ? super U> uncaptured) {
```

File: src/java.base/share/classes/java/util/function/BiFunction.java [hunk 3]
```diff
@@ -69,4 +69,30 @@ default <V> BiFunction<T, U, V> andThen(Function<? super R, ? extends V> after)
...
+     * var chained = BiFunction.of(String::repeat)     // BiFunction<String, Integer, String>
+     *                   .andThen(String::length);     // Function<String, Integer>
+     * }
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the first argument to the function
+     * @param <U> the type of the second argument to the function
+     * @param <R> the type of the result of the function
+     */
+    static <T, U, R> BiFunction<T, U, R> of(BiFunction<T, U, R> uncaptured) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/function/BiFunction.java:94]:
```suggestion
    static <T, U, R> BiFunction<T, U, R> of(BiFunction<? super T, ? super U, ? extends R> uncaptured) {
```

File: src/java.base/share/classes/java/util/function/BiPredicate.java [hunk 4]
```diff
@@ -104,4 +104,29 @@ default BiPredicate<T, U> or(BiPredicate<? super T, ? super U> other) {
...
+     * // Fluent composition
+     * var composed = BiPredicate.of(String::endsWith)     // BiPredicate<String, String>
+     *                    .or(String::startsWith);         // BiPredicate<String, String>
+     *}
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the first argument to the predicate
+     * @param <U> the type of the second argument the predicate
+     */
+    static <T, U> BiPredicate<T, U> of(BiPredicate<T, U> uncaptured) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/function/BiPredicate.java:128]:
```suggestion
    static <T, U> BiPredicate<T, U> of(BiPredicate<? super T, ? super U> uncaptured) {
```

File: src/java.base/share/classes/java/util/function/Consumer.java [hunk 5]
```diff
@@ -64,4 +64,28 @@ default Consumer<T> andThen(Consumer<? super T> after) {
...
+     *
+     * // Fluent composition
+     * var composed = Consumer.of(list::addLast)       // Consumer<String>
+     *                    .andThen(System.out::print); // Consumer<String>
+     *}
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the input to the operation
+     */
+    static <T> Consumer<T> of(Consumer<T> uncaptured) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/function/Consumer.java:87]:
```suggestion
    static <T> Consumer<T> of(Consumer<? super T> uncaptured) {
```

File: src/java.base/share/classes/java/util/function/Function.java [hunk 6]
```diff
@@ -97,4 +97,29 @@ default <V> Function<T, V> andThen(Function<? super R, ? extends V> after) {
...
+     * // Fluent composition
+     * var chained = Function.of(String::length)       // Function<String, Integer>
+     *                   .andThen(Integer::byteValue); // Function<String, Byte>
+     * }
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the input to the function
+     * @param <R> the type of the result of the function
+     */
+    static <T, R> Function<T, R> of(Function<T, R> uncaptured) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/function/Function.java:121]:
```suggestion
    static <T, R> Function<T, R> of(Function<? super T, ? extends R> uncaptured) {
```

File: src/java.base/share/classes/java/util/function/Predicate.java [hunk 7]
```diff
@@ -137,4 +137,28 @@ static <T> Predicate<T> not(Predicate<? super T> target) {
...
+     *
+     * // Fluent composition
+     * var composed = Predicate.of(String::isEmpty)
+     *                   .or(s -> s.startsWith("*")); // Predicate<String>
+     * }
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the input to the predicate
+     */
+    static <T> Predicate<T> of(Predicate<T> uncaptured) {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/function/Predicate.java:160]:
```suggestion
    static <T> Predicate<T> of(Predicate<? super T> uncaptured) {
```

**Review by ExE-Boss — COMMENTED** [review]
The factory methods should probably use bounded wildcards.

**Comment by minborg** [comment]
Maybe we should add a section to the java docs where we explain the implication if using a method reference and an overload is added later?

File: src/java.base/share/classes/java/util/function/BiConsumer.java (hunk 2 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/function/BiConsumer.java:104]: @ExE-Boss This and all other "enhancements" hardly compile ;-)

File: src/java.base/share/classes/java/util/function/BiConsumer.java [hunk 8]
```diff
@@ -72,4 +72,37 @@ default BiConsumer<T, U> andThen(BiConsumer<? super T, ? super U> after) {
             after.accept(l, r);
         };
     }
+
+    /**
+     * {@return a representation of the provided {@code uncaptured} lambda or method reference
+     * in the form of a {@code BiConsumer}}
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/util/function/BiConsumer.java:78]: There's no need for the provided arguments to this and the other "of" methods to be a lambda or a method reference expression, although it is true all that the methods are useful especially for such cases.

File: src/java.base/share/classes/java/util/function/BiConsumer.java [hunk 9]
```diff
@@ -72,4 +72,37 @@ default BiConsumer<T, U> andThen(BiConsumer<? super T, ? super U> after) {
...
+     *                    .andThen(this::toLogger);  // BiConsumer<Long, String>
+     *
+     * }
+     *
+     * @param uncaptured to capture
+     * @param <T> the type of the first argument to the operation
+     * @param <U> the type of the second argument to the operation
+     */
+    static <T, U> BiConsumer<T, U> of(BiConsumer<T, U> uncaptured) {
+        return Objects.requireNonNull(uncaptured);
```
Comment (rgiulietti) [inline src/java.base/share/classes/java/util/function/BiConsumer.java:105]: This does more than just capturing: not sure if the `null` check is generally useful here and elsewhere.

File: src/java.base/share/classes/java/util/function/BiConsumer.java (hunk 8 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/function/BiConsumer.java:78]: Also, is "capturing" the right term? IIRC, the JLS speaks about capture conversions for wildcards, but I'm not sure that the JLS concept also applies here.

File: src/java.base/share/classes/java/util/function/BiConsumer.java (hunk 9 above)
Comment (minborg) [reply src/java.base/share/classes/java/util/function/BiConsumer.java]: I have added a `@throws NullPointerException if source is null` clause where needed. I think there is no valid use case for providing null as a parameter and so I believe this is a desired property of the contemplated methods. However, there might be contrary arguments that I have not seen?

File: src/java.base/share/classes/java/util/function/UnaryOperator.java (hunk 1 above)
Comment (liach) [reply src/java.base/share/classes/java/util/function/UnaryOperator.java:68]: Just curious, do we have any idea how extensively `UnaryOperator` is used as a parameter type in user code compared to plain `Function<T, T>`? If it's not preferred over `Function<T, T>`, this new method might not be very useful.

Comment (minborg) [reply src/java.base/share/classes/java/util/function/UnaryOperator.java:68]: Generally speaking, there are 1,594 usages of `Function` and 208 usages of `UnaryFunction` in the JDK itself. More specifically, there are 8 usages of `Function::andThen` and zero use of `UnaryOperator::andThen` in the JDK.

Comment (minborg) [reply src/java.base/share/classes/java/util/function/UnaryOperator.java:68]: I've run some corpus analysis and this came up:

```
pminborg@pminborg-mac open % cat ~/Downloads/experiment-2.log | grep Function.andThen | wc -l
    6689
pminborg@pminborg-mac open % cat ~/Downloads/experiment-2.log | grep UnaryOperator.andThen | wc -l
     153
```

**Comment by mcimadamore** [comment]
I'm a little concerned that these methods will prove less useful than it seems (or _differently_ useful). That is, the main reason for adding them would be to allow fluent composition of operators, functions, etc. But this functional composition has limitation: since the target type only applies to the _last_ call in a chain, all the intermediate call will be type-checked without a target type. This means that any attempt to use implicit lambdas, or inexact method references in the intermediate links or the composition chain will result in inference errors and will, in my opinion, provide a not very smooth user experience.

So, if we exclude method chaining, the other reason as to why the new factories might be useful is that they allow developers to write slightly more succinct code, especially when combined with `var`. E.g. 

```
BinaryOperator<Integer> bop = Integer::sum
```

vs:

```
var bop = BinaryOperator.of(Integer::sum)
```

(of course the comparison gets more and more favorable for the latter the more type arguments appear in the functional interface type used in the LHS).

What about the vastly more common case where a functional interface is accepted as parameter to a Java method? Well, here it doesn't feel like the new factories add that much because:

* if no composition is needed, then target-type inference does its job just fine (e.g. just pass a lambda, or a method reference to the method)
* if composition is required, then, for the reasons above, the fluent style is hit and miss, and lack of target-typing support across the _entire_ method chain will result in inference errors.

So, I suspect I'm not 100% sold on the usefulness of these new factories.

**Comment by mcimadamore** [comment]
With my language hat on, there are two things that are problematic here:
1. lack of target-typing propagation can cause some use-site to be less fluent (as they have to break up different higher-order method calls)
2. the overload resolution logic for higher-order methods is sometimes overly conservative - e.g. it will refuse to look at the body of a lambda expression with implicit parameter types if there's two possible target types, even though the two targets agree on the inferred parameter types (this problem is described in the third part of https://openjdk.org/jeps/302).

Now, addressing (1) require big changes to the inference scheme (it is especially problematic if a "yet to be inferred" type-variable acts as a receiver for the subsequent method in the chain).

But addressing (2) is far more doable, and should lead to a nice usability boost in a lot of situations where the language currently says "sorry, ambiguous". And it feels this is the main motivation for adding the new APIs in the first place. If so, perhaps considering a language change/enhancement could provide better return on investment?

**Comment by minborg** [comment]
As much as I liked this idea from the beginning, I am getting increasingly skeptical. By introducing methods like these, we risk code compatibility over time and this feature would instead turn into a straight jacket. The objectives of this PR can be reached by language changes rather than the library approach attempted here. It would, for example, be possible to consider adding diamond operator/type inference in casts. So rather than `Function.of(String::isEmpty) `, one might write `(Function<>)String::isEmpty` which will be inferred to `(Function<String, Boolean>)String::isEmpty`.

**Pull request closed without merging**